C API: Burn `ConsolidateBlocks`

raynelfss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary
Add `qk_transpiler_pass_standalone_consolidate_blocks` to expose a version of the `ConsolidateBlocks` pass to C. The pass runs on an instance of `CircuitData` and an optional `Target`.

### Pre-requisites:
- [x] #14748
- [x] #14776
- [x] #14873

### Details and comments
~~Built on top of ~~#14748~~ and ~~#14776~~ needs to be rebased after it merges.~~ **Rebased!**
Closes #14447

mtreinish commented on crates/cext/src/transpiler/passes/consolidate_blocks.rs:
```diff
@@ -0,0 +1,185 @@
+
+/// Helper function that extracts the decomposer and basis gate directly from the [Target].
+#[inline]
+fn get_decomposer_and_basis_gate(
```
This should be integrated into the internal rust API. We will need a rust entrypoint that has the signature of the C API to call from the `transpile()` function. The C API should be doing appreciable logic like this it should just mirror the request to the rust api and handle basic type conversion from the limitations of the FFI boundary to something more rust native.

raynelfss committed:
Add: C FFI for `ConsolidateBlocks`
- Add `qk_transpiler_pass_standalone_consolidate_blocks` to expose a version of the `ConsolidateBlocks` pass to C. The pass runs on an instance of `CircuitData` and an optional `Target`.

raynelfss committed:
Chore: Add tests and fix formatting.

raynelfss committed:
Fix: C test deallocates

raynelfss committed:
Fix: Add missing void in function signature for c tests,

raynelfss committed:
Fix: Free circuit instruction instance afterwards.

raynelfss commented:
I've been trying to debug the failure on Windows happening here. I can't put my finger on exactly what is causing it to fail, but whenever we provide no target, which defaults `TwoQubitBasisDecomposer` with a CX gate, when we perform this call:
https://github.com/Qiskit/qiskit/blob/7547786a49da337072764752b629e20c18bc9a72/crates/transpiler/src/passes/consolidate_blocks.rs#L228
This is the last thing I can track within the pass, and the last line I can track within our crates is at the point where we call this line in the synthesis crate:
https://github.com/Qiskit/qiskit/blob/7547786a49da337072764752b629e20c18bc9a72/crates/synthesis/src/two_qubit_decompose.rs#L257
This is where Windows will segfault. I'm not sure as to why this is happening, though. Can anyone else try running these tests on their local machine on Windows and see if the segmentation fault happens?

eliarbel commented:
@raynelfss FWIW I have this standalone reproducer: 

```rust
use faer::mat;

use num_complex::Complex64;

#[inline(always)]
pub const fn c64(re: f64, im: f64) -> Complex64 {
    Complex64::new(re, im)
}
fn main() {
    let val = 0.9999999999999998;

    // Define the 4x4 complex matrix
    let matrix = mat![
        [c64(0.0, 0.0), c64(0.0, 0.0), c64(0.0, 0.0), c64(0.0, val)],
        [c64(0.0, 0.0), c64(0.0, 0.0), c64(0.0, -val), c64(0.0, 0.0)],
        [c64(0.0, 0.0), c64(0.0, -val), c64(0.0, 0.0), c64(0.0, 0.0)],
        [c64(0.0, val), c64(0.0, 0.0), c64(0.0, 0.0), c64(0.0, 0.0)],
    ];

    // Compute eigenvalues
    let result = matrix.complex_eigenvalues();
    println!("{:?}", result);

}
```

Which runs fine on Linux but results in this on Windows:
```text
thread 'main' has overflowed its stack
error: process didn't exit successfully: `target\debug\crash.exe` (exit code: 0xc00000fd, STATUS_STACK_OVERFLOW)
```
So obviously it's something with `faer` but this is how far I got at this point...

raynelfss committed:
FIx: Add missing `qk_count_ops_free` to Ctests

raynelfss committed:
Fix: Modify the Circuit in place.

raynelfss committed:
Docs: Add release note

raynelfss committed:
Chore: Add test with target

raynelfss committed:
Merge branch 'main' into burn-consolidate-blocks

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into burn-consolidate-blocks

raynelfss committed:
Fix: Incorrect formatting on c tests.

kevinhartman commented on crates/cext/src/transpiler/passes/consolidate_blocks.rs:
```diff
@@ -0,0 +1,72 @@
+            Some(const_ptr_as_ref(target))
+        }
+    };
+    let approximation_degree = if approximation_degree.is_nan() {
```
Do we expect to get a `NaN` here? The docs indicate that the value should be in the range `[0.0, 1.0]`.

raynelfss replied:
The `NaN` should specify the default option (which is 1.0). I'll mention this in the docs.

raynelfss replied:
Mentioned in docstring as of 12c94e07a5e8bd0b4a2d321f9d3d7c2e0821f58b

kevinhartman commented on test/c/test_consolidate_blocks.c:
```diff
@@ -0,0 +1,297 @@
+#include <stdio.h>
+#include <string.h>
+
+bool args_cmp(uint32_t *args1, uint32_t args1_len, uint32_t *args2, uint32_t args2_len);
```
Is there a reason to pre-declare this only to define it later in the same file?

raynelfss replied:
It's mostly just C syntax, I didn't want to include the helper functions at the top of the file. I can change this if it's not preferred.

kevinhartman replied:
Yeah that's what I was getting at. If you define it instead of pre-declare then we only ever have to update the signature in a single place. I don't mind though if you prefer it at the end of the file / want to keep it as is.

kevinhartman commented on test/c/test_consolidate_blocks.c:
```diff
@@ -0,0 +1,297 @@
+    // Run pass, without Target
+    qk_transpiler_pass_standalone_consolidate_blocks(circuit, NULL, 1.0, true);
+
+    QkOpCounts counts = qk_circuit_count_ops(circuit);
```
Beyond the scope of this PR, but I wonder why we return a struct from this instead of using the pattern of letting the user bring their own memory, accepting an "out" parameter, like we do for `qk_circuit_get_instruction`.

kevinhartman commented on test/c/test_consolidate_blocks.c:
```diff
@@ -0,0 +1,297 @@
+        goto cleanup;
+    }
+
+    QkCircuitInstruction *instruction = malloc(sizeof(QkCircuitInstruction));
```
This is currently a memory leak since we're not free-ing the `malloc` of the `QkCircuitInstruction` itself.

The `qk_circuit_instruction_clear` method only clears the internals. As a general rule, if you manually `malloc` something in the client code, you need to also manually `free` it. That's because Qiskit can't assume what allocator the user used when creating the memory for something it has been passed. 

The simplest fix here is to either call `free` after `qk_circuit_instruction_clear` under the `clean_instr` label, or just do a stack allocation instead of using the heap on this line.

```
QkCircuitInstruction instruction;
qk_circuit_get_instruction(circuit, 0, &instruction);
```

This would be the more-expected approach since there's no real need for this memory to be on the heap, and it'll go away automatically when this stack frame drops on function exit.

raynelfss replied:
Addressed by 12c94e07a5e8bd0b4a2d321f9d3d7c2e0821f58b

raynelfss replied:
Pivoted to calling free on d642f5c6202453a6e56a3223a8518623159a532a

kevinhartman replied:
Your issue with the first patch was that you need to actually allocate the struct on the stack, not just a pointer. Note how in my original suggestion above there is no `*` beside `instruction`. That's important since otherwise we're only allocating a pointer on the stack.

raynelfss replied:
Re-redo stack allocation aa3e7b074f67c94aed27192aa4dea2675b516199

kevinhartman commented on test/c/test_consolidate_blocks.c:
```diff
@@ -0,0 +1,297 @@
+        goto cleanup;
+    }
+
+    QkCircuitInstruction *unitary = malloc(sizeof(QkCircuitInstruction));
```
Same issue here with not having a call to `free`. Would be best to just do a stack allocation for `QkCircuitInstruction`.

raynelfss replied:
Addressed by 12c94e07a5e8bd0b4a2d321f9d3d7c2e0821f58b

kevinhartman commented on test/c/test_consolidate_blocks.c:
```diff
@@ -0,0 +1,297 @@
+    // Run the pass
+    qk_transpiler_pass_standalone_consolidate_blocks(qc, target, 1.0, true);
+
+    QkOpCounts counts = qk_circuit_count_ops(qc);
```
Would there also be any point in testing the correctness of the result here? I'm sure we exercise that already via Python, but just curious if it's a goal for us to validate it in Rust.

raynelfss replied:
I think we mostly focused on checking whether it does what it's supposed to, and we're leaving the correctness checks in Python for now because of how limited our testing is in Rust.

kevinhartman commented on crates/cext/Cargo.toml:
```diff
@@ -21,6 +21,7 @@ num-complex.workspace = true
 qiskit-quantum-info.workspace = true
 qiskit-circuit.workspace = true
 qiskit-transpiler.workspace = true
+qiskit-synthesis.workspace = true
```
Do we need this?

raynelfss replied:
I don't think so, this is possibly leftover code from when we had the decomposer extraction be performed from C.

raynelfss replied:
Removed by 12c94e07a5e8bd0b4a2d321f9d3d7c2e0821f58b

kevinhartman reviewed (commented):
Looks mostly good to me, just a few comments.

raynelfss committed:
Fix: Address review comments.
- Quit using malloc in favor of stack allocations.
- Remove dependency on synthesis module from `cext`.
- Add mention of `NaN` on docstring.

raynelfss committed:
Fix: Pivot to freeing the allocated space.

raynelfss committed:
Merge branch 'main' into burn-consolidate-blocks

raynelfss committed:
Fix: Adapt to #14930

raynelfss committed:
Fix: Redo stack approach correctly

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into burn-consolidate-blocks

kevinhartman reviewed (approved):
LGTM, thanks for the changes!
